controller/object_store_controller.go

```diff
@@ -0,0 +1,1150 @@
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil {
+		return errors.Wrapf(err, "failed to find pvc %v", genPVCName(store))
+	} else if err = osc.checkPVC(pvc); err != nil {
+		logrus.Errorf("Object Store running but PVC not bound")
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		return errors.Wrapf(err, "failed to find volume %v", genPVName(store))
+	} else if err = osc.checkVolume(vol); err != nil {
+		logrus.Errorf("Object Store running but Volume not ready")
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		return errors.Wrapf(err, "failed to find PV %v", genPVName(store))
+	} else if err = osc.checkPV(pv); err != nil {
+		logrus.Errorf("Object Store running but PV not ready")
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return errors.Wrap(err, "failed find Deployment to stop")
+	} else if (*dpl).Spec.Replicas != nil && *((*dpl).Spec.Replicas) != 0 {
+		(*dpl).Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	logrus.Infof("object store %v is now stopped", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		logrus.Infof("object store %v is now starting", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// remove finalizer and wait for dependent resources to be deleted
+	if len(store.ObjectMeta.Finalizers) != 0 {
+		return osc.ds.RemoveFinalizerForObjectStore(store)
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetDeployment(store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	return nil
+}
```

Comment:
The function `handleTerminating` currently remove the finalizer immediately and doesn't wait for the dependent resource. Suggesting to change it to:
```suggestion
func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore, key string) (err error) {
	// wait for dependent resources to be deleted then remove finalizer
	_, err = osc.ds.GetService(osc.namespace, store.Name)
	if err == nil {
		osc.queue.AddRateLimited(key)
		return nil
	}
	if !datastore.ErrorIsNotFound(err) {
		return err
	}

	_, err = osc.ds.GetDeployment(store.Name)
	if err == nil {
		osc.queue.AddRateLimited(key)
		return nil
	}
	if !datastore.ErrorIsNotFound(err) {
		return err
	}

	_, err = osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
	if err == nil {
		osc.queue.AddRateLimited(key)
		return nil
	}
	if !datastore.ErrorIsNotFound(err) {
		return err
	}

	_, err = osc.ds.GetPersistentVolume(genPVName(store))
	if err == nil {
		osc.queue.AddRateLimited(key)
		return nil
	}
	if !datastore.ErrorIsNotFound(err) {
		return err
	}

	_, err = osc.ds.GetVolume(genPVName(store))
	if err == nil {
		osc.queue.AddRateLimited(key)
		return nil
	}
	if !datastore.ErrorIsNotFound(err) {
		return err
	}

	return osc.ds.RemoveFinalizerForObjectStore(store)
}

```

Reply:
I wonder if we should get rid of the finalizer alltogether.
If you read the docs:
https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/#how-finalizers-work
https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/
https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents
it's not abundantly clear, but what happens is that as long as the finalizer remains, the object store isn't deleted - it just has a deletion timestamp. Therefore there is no garbage collection happening. To allow the object store to be deleted, the finalizer must be removed. Only then will the garbage collection delete the dependent resources.
I'm not sure there is much benefit in watching the resources either, since a) we don't delete them anyways, they are cleaned up by the K8s API and the CSI driver and b) we don't really want to re-queue resource events when the object store might already be deleted as this controller then has nothing to work with.

The alternative would be to keep the finalizer, remove the owner references and take care of cleaning up the resources in the controller. That would be quite a lot of code for very little benefit as far as I can tell.

Reply:
I see the point about K8s garbage collector doesn't clean up the dependent resources unless ObjectStore's finalizer is removed. This is the behavior when users delete the ObjectStore in background cascading mode. But if the users delete foreground cascading deletion mode, k8s will wait for dependent resources to be removed first. 

I think the general convention inside Longhorn controllers currently is controllers always enforce the foreground cascading deletion that is:
1. Controller trigger deletion for all dependent resources
2. Controller wait for all dependent resources to be fully removed
3. Controller remove finalizer for the main resource 

The benefit of the current design is that if the main resource is deleted and recreated immediately, the controller doesn't mistakenly adopt the dependent resource of the old object and assign it to the new object. For example, when an ObjectStore is deleted and recreated with the same name quickly, ObjectStore controller doesn't reuse the PV/Volume of the old ObjectStore for the new ObjectStore.

I do see that this approach does make the owner references mechanism a bit redundant. However, I can think of one usecase in which the owner references mechanism is still useful: in case the longhorn-manager pods are gone, users can delete and remove the finalizer of ObjectStore manually. K8s will handle the cleanup of dependent resources. 

In sum, WDYT about this approach:

1. Controller trigger deletion for all dependent resources
2. Controller wait for all dependent resources to be fully removed
3. Controller remove finalizer for the main resource 
4. Still keeping the the owner references mechanism

Reply:
> Why does the controller need to explicitly delete the dependency resources? Any extra concern about the foreground cascading deletion? The only exception (that cannot be handled by cascading deletion), PV, is already handled by the volume controller

If user deletes the ObjectStore using API, Longhorn manager can enforce the foreground cascading deletion. However, if user delete the volume ObjectStore using kubectl, Longhorn manager cannot enforce foreground cascading deletion. 

Right now, the volume controller inside Longhorn makes sure it deletes volume dependent resource before removing the finalizer. This means no matter how the user deletes the volume (API/kubectl), the deletion is foreground cascading. I think we should do the same for ObjectStore controller. WDYT?